pkg/auth/dependency/webapp/csp_middleware_test.go

```diff
@@ -0,0 +1,79 @@
+package webapp
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
+
+	"github.com/skygeario/skygear-server/pkg/core/config"
+)
+
+func TestCSPMiddleware(t *testing.T) {
+	Convey("CSPMiddleware", t, func() {
+		dummy := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {})
+		middleware := &CSPMiddleware{}
+		h := middleware.Handle(dummy)
+
+		Convey("no clients", func() {
+			w := httptest.NewRecorder()
+			r, _ := http.NewRequest("GET", "/", nil)
+			h.ServeHTTP(w, r)
+
+			So(w.Result().Header.Get("Content-Security-Policy"), ShouldEqual, "frame-ancestors 'self';")
+		})
+
+		Convey("one client", func() {
+			middleware.Clients = []config.OAuthClientConfiguration{
+				config.OAuthClientConfiguration{
+					"redirect_uris": []interface{}{
+						"https://example.com/path?q=1",
+					},
+				},
+			}
+			w := httptest.NewRecorder()
+			r, _ := http.NewRequest("GET", "/", nil)
+			h.ServeHTTP(w, r)
+
+			So(w.Result().Header.Get("Content-Security-Policy"), ShouldEqual, "frame-ancestors https://example.com 'self';")
+		})
+
+		Convey("more than one clients", func() {
+			middleware.Clients = []config.OAuthClientConfiguration{
+				config.OAuthClientConfiguration{
+					"redirect_uris": []interface{}{
+						"https://example.com/path?q=1",
+					},
+				},
+				config.OAuthClientConfiguration{
+					"redirect_uris": []interface{}{
+						"https://app.com/path?q=1",
+					},
+				},
+			}
+			w := httptest.NewRecorder()
+			r, _ := http.NewRequest("GET", "/", nil)
+			h.ServeHTTP(w, r)
+
+			So(w.Result().Header.Get("Content-Security-Policy"), ShouldEqual, "frame-ancestors https://example.com https://app.com 'self';")
+		})
+
+		Convey("include only https? redirect URIs", func() {
```

Comment:
We should not allow HTTP origin to embed us as iframe, since it is not secure.

Reply:
Sorry. `http` should only be allowed under this condition. https://github.com/SkygearIO/features/pull/425/files#diff-9176bd53b9cb2e7662018c59f2cdc01aR169

For example `localhost`, `127.0.0.1` and `skygear.localhost` are allowed but `abclocalhost` is not. Any objections?

Reply:
I think that's okay.